wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/RdsMultiAzDbClusterListProvider.java

```diff
@@ -0,0 +1,197 @@
+        conn.setNetworkTimeout(networkTimeoutExecutor, networkTimeout);
+      }
+    }
+  }
+
+  /**
+   * Get writer node ID.
+   *
+   * @param fetchWriterNodeResultSet A ResultSet of writer node query
+   * @return String The ID of a writer node
+   * @throws SQLException if errors occurred while retrieving the topology
+   */
+  private String processWriterNodeId(final ResultSet fetchWriterNodeResultSet) throws SQLException {
+    String writerNodeId = "";
+    while (fetchWriterNodeResultSet.next()) {
+      writerNodeId = fetchWriterNodeResultSet.getString(fetchWriterNodeQueryHeader);
+    }
+    return writerNodeId;
+  }
+
+  /**
+   * Form a list of hosts from the results of the topology query.
+   *
+   * @param topologyResultSet The results of the topology query
+   * @param writerNodeId The writer node ID
+   * @return a list of {@link HostSpec} objects representing
+   *     the topology that was returned by the
+   *     topology query. The list will be empty if the topology query returned an invalid topology
+   *     (no writer instance).
+   */
+  private List<HostSpec> processTopologyQueryResults(
+      final ResultSet topologyResultSet,
+      final String writerNodeId) throws SQLException {
+
+    final HashMap<String, HostSpec> hostMap = new HashMap<>();
+
+    // Data is result set is ordered by last updated time so the latest records go last.
+    // When adding hosts to a map, the newer records replace the older ones.
+    while (topologyResultSet.next()) {
+      final HostSpec host = createHost(topologyResultSet, writerNodeId);
+      hostMap.put(host.getHost(), host);
+    }
+
+    final List<HostSpec> hosts = new ArrayList<>();
+    final List<HostSpec> writers = new ArrayList<>();
+
+    for (final HostSpec host : hostMap.values()) {
+      if (host.getRole() != HostRole.WRITER) {
+        hosts.add(host);
+      } else {
+        writers.add(host);
+      }
+    }
+
+    int writerCount = writers.size();
+
+    if (writerCount == 0) {
+      LOGGER.severe(() -> Messages.get("RdsHostListProvider.invalidTopology"));
+      hosts.clear();
+    } else if (writerCount == 1) {
+      hosts.add(writers.get(0));
+    } else {
+      // Take the latest updated writer node as the current writer. All others will be ignored.
+      List<HostSpec> sortedWriters = writers.stream()
+          .sorted(Comparator.comparing(HostSpec::getLastUpdateTime).reversed())
+          .collect(Collectors.toList());
+      hosts.add(sortedWriters.get(0));
+    }
+
+    return hosts;
+  }
+
+  /**
+   * Creates an instance of HostSpec which captures details about a connectable host.
+   *
+   * @param resultSet the result set from querying the topology
+   * @return a {@link HostSpec} instance for a specific instance from the cluster
+   * @throws SQLException If unable to retrieve the hostName from the result set
+   */
+  private HostSpec createHost(final ResultSet resultSet, final String writerNodeId) throws SQLException {
```

Comment:
NIT: createHostSpec ?

Reply:
The `host` in the context is pretty much `HostSpec`. I'll keep it align with the default hostlist provider.